R/get_delay_estimate.R

```diff
@@ -54,47 +58,141 @@ get_delay_estimate <- function(reporting_triangle,
   # Produce a warning if the bottom left of a reporting trianglei s all 0s.
   if (isTRUE(.check_zeros_bottom_right(reporting_triangle))) {
     cli_warn(
-      message =
-        c(
-          "All entries in bottom right are 0. Are these true observations ",
-          "with zero reports, or are these unobserved? If the latter, ",
-          "replace with NA using the `replace_bottom_right_with_NA()` ",
-          "function."
-        )
+      message = c(
+        "All entries in bottom right are 0. Are these true observations ",
+        "with zero reports, or are these unobserved? If the latter, ",
+        "replace with NA using the `replace_bottom_right_with_NA()` ",
+        "function."
+      )
     )
   }
-  # Filter the reporting_triangle down to nrow = n_history_delay + 1,
-  # ncol = max_delay #nolint
+
+  # Filter the reporting_triangle down to relevant rows and columns
+  trunc_triangle <- .prepare_triangle(reporting_triangle, max_delay, n)
+
+  # Fill in missing values in the triangle
+  expectation <- .fill_triangle(trunc_triangle)
+
+  # Calculate probability mass function from filled triangle
+  pmf <- .calculate_pmf(expectation)
+
+  return(pmf)
+}
+
+#' Prepare the triangle by truncating and handling negative values
+#'
+#' @param reporting_triangle The original reporting triangle
+#' @param max_delay Maximum delay to consider
+#' @param n Number of reference times to use
+#' @return Prepared reporting triangle
+#' @noRd
+.prepare_triangle <- function(reporting_triangle, max_delay, n) {
   nr0 <- nrow(reporting_triangle)
   trunc_triangle <- reporting_triangle[(nr0 - n + 1):nr0, 1:(max_delay + 1)]
   rep_tri <- .handle_neg_vals(trunc_triangle)
+  return(rep_tri)
+}
+
+#' Fill in missing values in the reporting triangle
+#'
+#' @param rep_tri Prepared reporting triangle
+#' @return Matrix with imputed values for missing entries
+#' @noRd
+.fill_triangle <- function(rep_tri) {
   n_delays <- ncol(rep_tri)
   n_dates <- nrow(rep_tri)
-  mult_factor <- vector(length = max_delay - 1)
   expectation <- rep_tri
+
   # Find the column to start filling in
   start_col <- which(colSums(is.na(rep_tri)) > 0)[1]
 
   # Only fill in reporting triangle if it is incomplete
   if (!is.na(start_col)) {
```

Comment:
If you pass in a complete reporting matrix, won't `start_col` be NA, and this whole step of filling in the triangle should be skipped? That is why I had the outer `if(!is.na(start_col))`

Reply:
Ah I was thrown off by:  reporting_triangle as the input name can add back in.

Reply:
wait there is a is.na check here so this case is covered?

Reply:
Yes, sorry I think I read this as having got deleted before